OpenEdXMobile/src/main/java/org/edx/mobile/viewModel/CourseViewModel.kt

```diff
@@ -105,6 +117,82 @@ class CourseViewModel @Inject constructor(
         }
     }
 
+    fun getCourseData(
+        courseId: String,
+        courseComponentId: String?,
+        showProgress: Boolean = false,
+        swipeRefresh: Boolean = false,
+        coursesRequestType: CoursesRequestType = LIVE
+    ) {
+        viewModelScope.launch {
+            _swipeRefresh.postValue(swipeRefresh)
+            _showProgress.postValue(showProgress)
+            val courseComponentData: kotlin.Result<CourseComponent?> = when (coursesRequestType) {
+                CACHE -> {
+                    runCatching {
+                        courseRepository.getCourseComponentsFromCache(
+                            courseId,
+                            courseComponentId,
+                            Dispatchers.IO
+                        )
+                    }
+                }
+                STALE, LIVE -> {
+                    runCatching {
+                        courseRepository.getCourseStructure(
+                            courseId,
+                            coursesRequestType,
+                            Dispatchers.IO
+                        )
+                    }
+                }
+                else -> {
+                    throw Exception("Unknown Request Type: $coursesRequestType")
+                }
+            }
+            courseComponentData.onSuccess { courseComponent ->
+                _swipeRefresh.postValue(false)
+                _showProgress.postValue(false)
+                if (courseComponent == null) {
+                    if ((coursesRequestType == CACHE || coursesRequestType == STALE)) {
+                        getCourseData(
+                            courseId,
+                            courseComponentId,
+                            showProgress = showProgress,
+                            swipeRefresh = false,
+                            coursesRequestType = LIVE
+                        )
+                    }
+                } else {
+                    _courseComponent.postEvent(courseComponent)
+                    // if course data exist in persistable cache update the course data in background
+                    if (coursesRequestType == CACHE) {
```

Comment:
The stale call should be sent only after a Persistable/Server cache call. Right now, we will call it even if we get data from AppLevelCache.